### PR TITLE
add stateful checkpointer into tmaster

### PR DESCRIPTION
**Pull request by nlu90**
The tmaster stateful utils that issue checkpoint requests to stmgrs and tracking if the a global consistent checkpoint is done.

No logic changes in tmaster code, so won't affect any production performance now.

### CODE REVIEW

File: heron/tmaster/src/cpp/manager/tmaster.h [hunk 1]
```diff
@@ -37,6 +37,10 @@ class StatsInterface;
 class TMasterServer;
 class TMetricsCollector;
 
+typedef std::map<std::string, StMgrState*> StMgrMap;
+typedef StMgrMap::iterator StMgrMapIter;
+typedef StMgrMap::const_iterator StMgrMapConstIter;
```
Comment (billonahill) [inline heron/tmaster/src/cpp/manager/tmaster.h]: Why is `StMgrMapConstIter` needed? I don't see it used anywhere.

File: heron/tmaster/src/cpp/manager/stateful-checkpointer.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+namespace heron {
+namespace proto {
+class PhysicalPlan;
+}
+}
+
+namespace heron {
+namespace tmaster {
+
+class StatefulCheckpointer {
```
Comment (billonahill) [inline heron/tmaster/src/cpp/manager/stateful-checkpointer.h]: please add class-level docs describing what this class does and what it's responsibilities are.

File: heron/tmaster/src/cpp/manager/stateful-checkpointer.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  explicit StatefulCheckpointer(std::chrono::high_resolution_clock::time_point _tmaster_start_time);
+  virtual ~StatefulCheckpointer();
+  void RegisterNewPhysicalPlan(const proto::system::PhysicalPlan& _pplan);
+
+  void StartCheckpoint(const StMgrMap& _stmgrs);
+
+  // Called when a InstanceStateStored message is received
+  // Return true if this completes a globally consistent checkpoint
+  // for this _checkpoint_id
+  bool HandleInstanceStateStored(const std::string& _checkpoint_id,
```
Comment (billonahill) [inline heron/tmaster/src/cpp/manager/stateful-checkpointer.h]: who calls `HandleInstanceStateStored`? Why does this need to be public?

File: heron/tmaster/src/cpp/manager/stateful-checkpointer.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+#include <unordered_set>
+#include "network/network.h"
+#include "proto/tmaster.pb.h"
+#include "manager/tmaster.h"
+#include "basics/basics.h"
+
+
+namespace heron {
+namespace proto {
+class PhysicalPlan;
```
Comment (billonahill) [inline heron/tmaster/src/cpp/manager/stateful-checkpointer.h]: Why is this redefined here?

File: heron/tmaster/src/cpp/manager/stateful-checkpointer.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  : tmaster_start_time_(_tmaster_start_time) {
+  // do nothing
+}
+
+StatefulCheckpointer::~StatefulCheckpointer() { }
+
+sp_string StatefulCheckpointer::GenerateCheckpointId() {
+  // TODO(skukarni) Should we append any topology name/id stuff?
+  std::ostringstream tag;
+  tag << tmaster_start_time_.time_since_epoch().count()
```
Comment (billonahill) [inline heron/tmaster/src/cpp/manager/stateful-checkpointer.cpp]: How do we know that a checkpoint is valid for a given packing plan? Should we build this into the id or bind it some other way? We will need this to know if a checkpoint can be assigned to a redeployed topology with a different packing plan.

File: heron/tmaster/src/cpp/manager/tmaster.h (hunk 1 above)
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/tmaster.h]: removed. Will add it when used in the future code

File: heron/tmaster/src/cpp/manager/stateful-checkpointer.h (hunk 4 above)
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/stateful-checkpointer.h]: This is not redefining but forward declaration.  Details about forward declaration could be found here: https://stackoverflow.com/questions/4757565/what-are-forward-declarations-in-c

File: heron/tmaster/src/cpp/manager/stateful-checkpointer.h (hunk 2 above)
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/stateful-checkpointer.h]: added

File: heron/tmaster/src/cpp/manager/stateful-checkpointer.h (hunk 3 above)
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/stateful-checkpointer.h]: When tmaster received the InstanceStateStored message, it calls this method to notify the checkpointer. And also, checkpointer will let tmaster know if all states for the checkpoint_id are stored.

I'll update the comment

File: heron/tmaster/src/cpp/manager/stateful-checkpointer.cpp (hunk 5 above)
Comment (nlu90) [inline heron/tmaster/src/cpp/manager/stateful-checkpointer.cpp]: This is not a checkpointer's job. tmaster will be responsible for it.

**Comment by nlu90** [comment]
unit tests added

File: heron/tmaster/tests/cpp/server/stateful_checkpointer_unittest.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+#include "errors/modinit.h"
+#include "threads/modinit.h"
+#include "network/modinit.h"
+#include "config/topology-config-vars.h"
+#include "config/topology-config-helper.h"
+#include "config/physical-plan-helper.h"
+#include "manager/stateful-checkpointer.h"
+
+// Generate a dummy topology
+static heron::proto::api::Topology* GenerateDummyTopology(
```
Comment (billonahill) [inline heron/tmaster/tests/cpp/server/stateful_checkpointer_unittest.cpp]: I've seen this method in a number of test classes. If you have ideas for how to effectively share test code like this that would be awesome.

**Pull request merged**